sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnInvokers.java

```diff
@@ -36,4 +36,26 @@
   }
 
   private DoFnInvokers() {}
+
+  /**
+   * Tries to invoke setup on the given {@link DoFn}. If setup throws any exception, the given
+   * {@link DoFn} will be torn down immediately and the exception rethrown.
+   *
+   * <p>On success returns an {@link DoFnInvoker} for the given {@link DoFn}.
+   */
+  public static <InputT, OutputT> DoFnInvoker<InputT, OutputT> tryInvokeSetupFor(
+      DoFn<InputT, OutputT> fn) {
+    DoFnInvoker<InputT, OutputT> doFnInvoker = invokerFor(fn);
+    try {
+      doFnInvoker.invokeSetup();
+    } catch (Exception e) {
+      try {
+        doFnInvoker.invokeTeardown();
+      } catch (Exception ignored) {
+        // ignored, we already have an exception to throw
```

Comment:
Would it make sense to do `e.addSuppressed(ignored)`?

Reply:
Sure! Thanks for pointing me to that direction.